### PR TITLE
Proxy Agent Provisioning Status Timeout

### PR DESCRIPTION
**Pull request by shahneerali**

### CODE REVIEW

File: proxy_agent_shared/src/telemetry/event_logger.rs [hunk 1]
```diff
@@ -75,6 +81,45 @@ pub fn write_state_event(
         }
     }
 }
+
+pub fn provisioning_timeout(key: &str) -> String {
```
Comment (ZhidongPeng) [inline proxy_agent_shared/src/telemetry/event_logger.rs:85]: rename this function

Comment (ZhidongPeng) [reply proxy_agent_shared/src/telemetry/event_logger.rs:85]: need unit test for this function

Comment (ZhidongPeng) [reply proxy_agent_shared/src/telemetry/event_logger.rs:85]: move this function to another place since it is not related to event_logger.

File: proxy_agent_shared/src/telemetry/event_logger.rs [hunk 2]
```diff
@@ -75,6 +81,45 @@ pub fn write_state_event(
...
+    unsafe {
+        match CIRCUIT_BREAKER_MAP.get(key) {
+            Some(v) => {
```
Comment (ZhidongPeng) [inline proxy_agent_shared/src/telemetry/event_logger.rs:88]: please add comments here to explain how the state-machine transition works in the code. you will have something special to success, transition, fail keyword in your code.

**Review by ZhidongPeng — DISMISSED** [review]

File: proxy_agent_extension/src/service_main.rs [hunk 3]
```diff
@@ -209,13 +215,15 @@ fn extension_substatus(
...
     status: &mut StatusObj,
+    status_state_obj: &mut common::StatusState,
 ) {
     let proxy_agent_aggregate_status_obj =
         proxy_agent_aggregate_status_top_level.proxyAgentStatus;
 
     let proxy_agent_aggregate_status_file_version =
         proxy_agent_aggregate_status_obj.version.to_string();
     if proxy_agent_aggregate_status_file_version != *proxyagent_file_version_in_extension {
+        common::StatusState::update_state(status_state_obj, false);
```
Comment (ZhidongPeng) [inline proxy_agent_extension/src/service_main.rs:226]: you need use this return value to set the status

File: proxy_agent_extension/src/service_main.rs [hunk 4]
```diff
@@ -176,6 +181,7 @@ fn report_proxy_agent_aggregate_status(
                 "service_main",
                 constants::SERVICE_LOG_FILE,
             );
+            _ = common::StatusState::update_state(status_state_obj, false);
```
Comment (ZhidongPeng) [inline proxy_agent_extension/src/service_main.rs:184]: you need this return value to set the status

File: proxy_agent_extension/src/constants.rs [hunk 5]
```diff
@@ -42,4 +42,7 @@ pub const MIN_SUPPORTED_OS_BUILD: u32 = 17763;
 pub const EXIT_CODE_SERVICE_STOP_ERR: i32 = 14;
 pub const EXIT_CODE_UPDATE_TO_VERSION_ENV_VAR_NOTFOUND: i32 = 15;
 pub const STATE_KEY_READ_PROXY_AGENT_STATUS_FILE: &str = "ReadProxyAgentStatusFile";
-pub const STATE_KEY_FILE_VERSION: &str = "FileVersion";
\ No newline at end of file
+pub const STATE_KEY_FILE_VERSION: &str = "FileVersion";
+pub const PROVISIONING_STATUS_KEY: &str = "proxyagentstatus";
```
Comment (ZhidongPeng) [inline proxy_agent_extension/src/constants.rs:46]: clean up the unnecessary changes.

File: proxy_agent_extension/src/common.rs [hunk 6]
```diff
@@ -273,6 +273,63 @@ pub fn stop_event_logger() {
...
 }
 
+pub struct StatusState {
+    pub current_state: String,
+    pub consecutive_fail_count: u32,
+    pub consecutive_success_count: u32,
+    pub transition_to_error_threshold: u32,
+}
+
+const MAX_CONSECUTIVE_COUNT:u32 = 10000;
```
Comment (ZhidongPeng) [inline proxy_agent_extension/src/common.rs:283]: move it inside the impl StatusState?

File: proxy_agent_extension/src/common.rs [hunk 7]
```diff
@@ -506,4 +563,61 @@ mod tests {
...
+        status_state_obj.consecutive_success_count = 2;
+        super::StatusState::update_state(&mut status_state_obj, true);
+        assert_eq!(
+            status_state_obj.current_state,
+            constants::SUCCESS_STATUS.to_string()
+        );
+
+        // Case 3: Testing Transitioning to Error
+        status_state_obj.current_state = constants::TRANSITIONING_STATUS.to_string();
+        status_state_obj.consecutive_fail_count = 21;
```
Comment (ZhidongPeng) [inline proxy_agent_extension/src/common.rs:591]: 19?

File: proxy_agent_extension/src/common.rs [hunk 8]
```diff
@@ -506,4 +563,61 @@ mod tests {
...
+        status_state_obj.consecutive_fail_count = 2;
+        super::StatusState::update_state(&mut status_state_obj, false);
+        assert_eq!(
+            status_state_obj.current_state,
+            constants::TRANSITIONING_STATUS.to_string()
+        );
+
+        // Case 2: Testing Transitioning to Success
+        status_state_obj.current_state = constants::TRANSITIONING_STATUS.to_string();
+        status_state_obj.consecutive_success_count = 2;
```
Comment (ZhidongPeng) [inline proxy_agent_extension/src/common.rs:582]: set consecutive_fail_count = 2 instead

File: proxy_agent_extension/src/common.rs [hunk 9]
```diff
@@ -506,4 +563,61 @@ mod tests {
...
+        status_state_obj.consecutive_fail_count = 21;
+        super::StatusState::update_state(&mut status_state_obj, false);
+        assert_eq!(
+            status_state_obj.current_state,
+            constants::ERROR_STATUS.to_string()
+        );
+
+        // Case 4: Testing Error to Transitioning
+        status_state_obj.current_state = constants::ERROR_STATUS.to_string();
+        status_state_obj.consecutive_success_count = 2;
```
Comment (ZhidongPeng) [inline proxy_agent_extension/src/common.rs:600]: set consecutive_fail_count = 2 instead

File: proxy_agent_extension/src/common.rs [hunk 10]
```diff
@@ -506,4 +563,61 @@ mod tests {
...
     }
+
+    #[test]
+    fn test_StatusState() {
+        let mut status_state_obj = super::StatusState::new();
+
+        // Case 1: Testing Success to Transitioning 
+        status_state_obj.current_state = constants::SUCCESS_STATUS.to_string();
+        status_state_obj.consecutive_fail_count = 2;
+        super::StatusState::update_state(&mut status_state_obj, false);
```
Comment (ZhidongPeng) [inline proxy_agent_extension/src/common.rs:574]: let current_state = status_state_obj.update_state(false);

same as below

**Review by ZhidongPeng — CHANGES_REQUESTED** [review]

File: proxy_agent_extension/src/common.rs (hunk 7 above)
Comment (shahneerali) [reply proxy_agent_extension/src/common.rs:591]: shouldn't the fail count be greater than the threshold which is 20 to trigger an error state

File: proxy_agent_extension/src/common.rs [hunk 11]
```diff
@@ -506,4 +563,61 @@ mod tests {
...
         );
     }
+
+    #[test]
+    fn test_StatusState() {
+        let mut status_state_obj = super::StatusState::new();
+
+        // Case 1: Testing Success to Transitioning 
+        status_state_obj.current_state = constants::SUCCESS_STATUS.to_string();
+        status_state_obj.consecutive_fail_count = 2;
```
Comment (ZhidongPeng) [inline proxy_agent_extension/src/common.rs:573]: Set Success counter to 2 here

**Review by ZhidongPeng — DISMISSED** [review]

**Review by ZhidongPeng — APPROVED** [review]
